File: 2019-03-05-222/lb/222.java
public class CountNodes {
    boolean tag = true;
    int count = 0;//最后一层的节点数

    /**
     * 根据层数和最后一层的节点数来计算这颗树的节点数
     * @param root
     * @return
     */
    public int countNodes(TreeNode root) {
        if (root == null)
            return 0;
        int h=0;//高度
        TreeNode temp = root;
        while (temp != null){
            h++;
            temp = temp.left;
        }
        //求末尾层的叶子点
        countLeaf(1,h,root);
        //count计算所求叶子节点的数量
        //可以计算左右叶子节点，所以需要除于2，这样可以直接根据层数直接判断
        return count+ (int)Math.pow(2,h-1)-1;
    }
    public void countLeaf(int cur,int h,TreeNode node){
        if (tag){
            if (cur == h-1 && node == null){
                tag = false;
                return;
            }
            if (cur == h && node != null) {
                count++;
                return;
            }
            if (node == null){
                return;
            }
            cur++;
            countLeaf(cur,h,node.left);
            countLeaf(cur,h,node.right);
        }
    }
}

File: Year-2019/Mar/2019-03-05-222/watercow/222.py
#!/usr/bin/env python  
#-*- coding:utf-8 _*-  
""" 
@author:watercow 
@license: Apache Licence 
@file: test.py 
@time: 2019/03/05
@site:  
@software: PyCharm 

# code is far away from bugs with the god animal protecting
    I love animals. They taste delicious.
              ┏┓      ┏┓
            ┏┛┻━━━┛┻┓
            ┃      ☃      ┃
            ┃  ┳┛  ┗┳  ┃
            ┃      ┻      ┃
            ┗━┓      ┏━┛
                ┃      ┗━━━┓
                ┃  神兽保佑    ┣┓
                ┃　永无BUG！   ┏┛
                ┗┓┓┏━┳┓┏┛
                  ┃┫┫  ┃┫┫
                  ┗┻┛  ┗┻┛ 
"""
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def __init__(self):
        self.count = 0
        self.flag = False
        self.LevelLastnum = 0 # 最后一层的节点数

    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        r = root
        level_count = 0 # 记录层数
        while r:
            level_count += 1
            r = r.left

        num1 = 2**(level_count-1) - 1 # 到level_count-1层的节点数

        self.preorder(1, level_count, root) # root层记为1而不是0

        return self.count/2 + num1

    def preorder(self, cur_level, level_count, node):
        if not self.flag:
            if not node and cur_level == level_count:
                self.flag = True
                return

            if cur_level == level_count + 1:
                self.count += 1
                return

            self.preorder(cur_level + 1, level_count, node.left)
            self.preorder(cur_level + 1, level_count, node.right)